Builder/TreeBuilder.php

```diff
@@ -0,0 +1,246 @@
+<?php
+
+namespace Becklyn\RouteTreeBundle\Builder;
+
+use Becklyn\RouteTreeBundle\Exception\InvalidRouteTreeException;
+use Becklyn\RouteTreeBundle\Tree\Node;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+use Symfony\Component\PropertyAccess\PropertyAccessor;
+use Symfony\Component\Routing\Route;
+use Symfony\Component\Routing\RouteCollection;
+
+
+/**
+ * Builds a tree from a list of routes
+ */
+class TreeBuilder
+{
+    const ROUTE_OPTIONS_KEY = "page_tree";
+
+    /**
+     * All configuration options.
+     *
+     * Mapping of routing configuration to the properties of the Node
+     * ["config-option" => "propertyPath"]
+     *
+     * @var string[]
+     */
+    private static $configurationOptions = [
+        "title" => "title",
+        "hidden" => "hidden",
+        "separator" => "separator",
+        "parameters" => "parameters",
+        "security" => "security",
+        "parent" => "parentRoute",
+    ];
+
+
+    /**
+     * @var PropertyAccessor
+     */
+    private $propertyAccessor;
+
+
+    /**
+     * @var ParametersGenerator
+     */
+    private $parametersGenerator;
+
+
+
+    /**
+     * @param ParametersGenerator $parametersGenerator
+     */
+    public function __construct (ParametersGenerator $parametersGenerator)
+    {
+        $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
+        $this->parametersGenerator = $parametersGenerator;
+    }
+
+
+
+    /**
+     * Builds the route tree
+     *
+     * @param Route[]|RouteCollection $routeCollection
+     *
+     * @return Node[] the array is indexed by route name
+     */
+    public function buildTree ($routeCollection)
+    {
+        $relevantRoutes = $this->calculateRelevantRoutes($routeCollection);
+        $nodes = $this->generateNodesFromRoutes($routeCollection, $relevantRoutes);
+        $nodes = $this->linkNodeHierarchy($nodes);
+
+        // needs to be after the hierarchy has been set up
+        return $this->calculateAllParameters($nodes);
+    }
+
+
+
+    /**
+     * Calculates a list of which routes should be included in the tree
+     *
+     * @param Route[]|RouteCollection $routeCollection
+     *
+     * @return array of format ["route" => (bool) $includeInTree]
+     * @throws InvalidRouteTreeException
+     */
+    private function calculateRelevantRoutes ($routeCollection)
+    {
+        $routeIndex = [];
+
+        foreach ($routeCollection as $routeName => $route)
+        {
+            $routeIndex[$routeName] = false;
+        }
+
+        foreach ($routeCollection as $routeName => $route)
```

Comment:
Because we need to ensure that **all** keys are set. As we access the parent node in line 125, we can't be sure if it was visited before. The nodes don't need to be order topologically.